### PR TITLE
Lint using Ruff

### LINKED ISSUE #6325: Tracking issue - linting/code quality using Ruff
This is a tracking issue for the introduction of code quality linting to this repo. Linting can provide the following benefits

- correctness - identify bugs, dead code, inconsistencies, performance issues, antipatterns in code
- consistency - promote a consistent code style to improve readability

I propose an incremental approach where lints are added one at a time, and only where they add clear value.

I propose using [ruff](https://github.com/charliermarsh/ruff). The alternative would be to use flake8. Compared to flake8, ruff
- is faster (by a couple of orders of magnitude)
- can automatically fix code in many cases
- has no dependencies

Ruff is particularly suited for use as a pre-commit hook, due to its speed.

see-

- [x] #6321 
- [x]  #6319 https://pypi.org/project/flake8-simplify/0.19.3/
- [ ]  #6320 
- [x] #6315 
- [x] #6317 
- [x] #6345
- [x] #6350 
- [x] #6351 

---------------------------------------------------------------------------------------------

Current "approved" rust.toml (with atleast one approval)
```
select = [
    "C4",
    "E713",
    "SIM2",  # simplify boolean comparisons
    "SIM110",  # convert loop to 'any'
    "SIM111",  # convert loop to 'all'
    "SIM109",  # use a tuple for multiple comparisons
    "SIM118",  # use 'key in dict'
    "PGH003",  # forbid blanket 'type: ignore' comments
]
```

### PR DESCRIPTION
**Pull request by danieleades**
closes #6325

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -28,3 +28,15 @@ repos:
...
         files: \.(html|md|yml|yaml)
         args: [--prose-wrap=preserve]
+
+  - repo: https://github.com/charliermarsh/ruff-pre-commit
+<<<<<<< HEAD
+    rev: v0.0.215
+=======
+    rev: v0.0.208
+>>>>>>> 165b683a (use ruff to lint and fix E713)
+    hooks:
```
Comment (MridulS) [inline .pre-commit-config.yaml:38]: There is a not fixed merge conflict here.

**Comment by danieleades** [comment]
@MridulS any other actions left here to get this merged?

**Comment by danieleades** [comment]
@MridulS this PR seems to be stuck in purgatory. Anything I can do to move this along?

**Review by dschult — APPROVED** [review]
It looks like 7 out of the 8 lint PRs listed in #6325 have been approved by @mridul and the 8th has been postponed because it will be included when version numbers increase.

But since those approvals, this PR was updated (due to updated versions of the tools). Do we need to go through all the changes again to see what is new that might have been thrown in there? Maybe it would have been better to leave the original commits unsquashed so we could tell what happened in the more recent updates. If that is possible, it might help the reviewing process. 

The only other concern is that this might cause many open PRs to have conflicts with  these changes. But I think we are Ok with putting up with that -- and reviewers can resolve those conflicts to help the submitters. 

Any other thoughts?  @MridulS you have done the most work on this. If you think it is ready go ahead and merge...  I've given my approval here but it is contingent on @MridulS 's approval.   :)

**Comment by MridulS** [comment]
Thanks for the review @dschult!

Thanks @danieleades for getting this together, it's time to ruff-ify now :)

**Comment by MridulS** [comment]
@dschult we can check ruff "in production" I guess. If issues do pop up we can revert this PR.

I'll also add the main branch commit to git-blame-ignore so this doesn't pollute git blame.

**Pull request merged**